File: backend/Java/src/cn/easyar/webar/WebAR.java
package cn.easyar.webar;

import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.security.MessageDigest;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.SortedMap;
import java.util.TreeMap;

public class WebAR {
    private String cloudKey;
    private String cloudSecret;
    private String cloudUrl;

    private OkHttpClient http;

    public WebAR(String cloudKey, String cloudSecret, String cloudUrl) {
        this.cloudKey = cloudKey;
        this.cloudSecret = cloudSecret;
        this.cloudUrl = cloudUrl;

        this.http = new OkHttpClient();
    }

    /**
     * 将图片数据签名，并发送到识别服务器
     * @param image 图片的base64数据
     * @return
     * @throws IOException
     */
    public ResultInfo recognize(String image) throws IOException {
        SortedMap<String, String> params = new TreeMap<>();
        params.put("image", image);
        params.put("timestamp", String.valueOf(System.currentTimeMillis()));
        params.put("appKey", this.cloudKey);
        params.put("signature", this.getSign(params));

        ObjectMapper mapper = new ObjectMapper();
        String str = mapper.writeValueAsString(params);

        return mapper.readValue(this.send(str), ResultInfo.class);
    }

    private String getSign(SortedMap<String, String> params) {
        StringBuilder builder = new StringBuilder();
        params.forEach((key, value) -> builder.append(key + value));

        return this.sha256(builder.toString() + this.cloudSecret);
    }

    private String sha256(String str) {
        StringBuilder builder = new StringBuilder();
        try {
            MessageDigest digest = MessageDigest.getInstance("sha-256");
            digest.update(str.getBytes());
            byte[] bytes = digest.digest();

            for (byte b: bytes) {
                String hex = Integer.toHexString(b & 0XFF);
                builder.append(hex.length() == 1 ? "0" + hex : hex);
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        return builder.toString();
    }

    private String getUtcDate() {
        return ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT);
    }

    private byte[] send(String json) throws IOException {
        RequestBody body = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), json);
        Request request = new Request.Builder()
                .url(this.cloudUrl)
                .post(body)
                .build();

        try (Response response = this.http.newCall(request).execute()) {
            return response.body().bytes();
        }
    }
}

File: backend/C#/cn/easyar/webar/WebAR.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Net.Http;
using System.Net.Http.Headers;
using Newtonsoft.Json;


namespace cn.easyar.webar {
    class WebAR {
        private string cloudKey;
        private string cloudSecret;
        private string cloudUrl;

        public WebAR(string cloudKey, string cloudSecret, string cloudUrl) {
            this.cloudKey = cloudKey;
            this.cloudSecret = cloudSecret;
            this.cloudUrl = cloudUrl;
        }

        public ResultInfo Recognize(string image) {
            SortedDictionary<string, string> param = new SortedDictionary<string, string>();
            param.Add("image", image);
            param.Add("timestamp", this.getTimestamp());
            param.Add("appKey", this.cloudKey);
            param.Add("signature", this.getSign(param));

            string json = JsonConvert.SerializeObject(param);
            string str = this.send(json);

            return JsonConvert.DeserializeObject<ResultInfo>(str);
        }

        private string getTimestamp() {
            TimeSpan ts = DateTime.UtcNow - new DateTime(1970, 1, 1, 0, 0, 0);
            return Convert.ToInt64(ts.TotalMilliseconds).ToString();
        }

        private string getSign(SortedDictionary<string, string> param) {
            StringBuilder builder = new StringBuilder();
            foreach (KeyValuePair<string, string> item in param) {
                builder.Append(item.Key + item.Value);
            }

            return this.sha256(builder.ToString() + this.cloudSecret);
        }

        private string sha256(string str) {
            SHA256CryptoServiceProvider sha256 = new SHA256CryptoServiceProvider();
            byte[] bytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(str));

            StringBuilder builder = new StringBuilder();
            foreach (byte b in bytes) {
                builder.Append(b.ToString("x2"));
            }

            return builder.ToString();
        }

        private string send(string json) {
            string str = "";
            try {
                HttpClient client = new HttpClient();

                HttpContent content = new StringContent(json);
                content.Headers.ContentType = new MediaTypeHeaderValue("application/json");

                HttpResponseMessage obj = client.PostAsync(this.cloudUrl, content).Result;
                str = obj.Content.ReadAsStringAsync().Result;
            } catch (Exception e) {
                System.Console.WriteLine(e.Message);
            }

            return str;
        }
    }
}
